File: internal/escrow/service.go
package escrow

import (
	"context"
	"time"

	"github.com/karada/internal/lightning"
	"github.com/karada/pkg/crypto"
)

// Service handles escrow business logic
type Service struct {
	repo   *Repository
	ln     *lightning.Client
	expiry int64
}

// NewService creates a new escrow service
func NewService(repo *Repository, ln *lightning.Client, expiry int64) *Service {
	return &Service{
		repo:   repo,
		ln:     ln,
		expiry: expiry,
	}
}

// CreateEscrowRequest represents the request to create an escrow
type CreateEscrowRequest struct {
	AmountSats  int64  `json:"amount_sats"`
	Description string `json:"description"`
}

// CreateEscrowResponse represents the response with payment request
type CreateEscrowResponse struct {
	PaymentRequest string `json:"payment_request"`
	PaymentHash    string `json:"payment_hash"`
	Preimage       string `json:"preimage"`
	AmountSats     int64  `json:"amount_sats"`
	Status         string `json:"status"`
	Expiry         int64  `json:"expiry"`
}

// CreateEscrow creates a new HODL invoice escrow
func (s *Service) CreateEscrow(ctx context.Context, req *CreateEscrowRequest) (*CreateEscrowResponse, error) {
	// Generate preimage and payment hash
	preimage, err := crypto.GeneratePreimage()
	if err != nil {
		return nil, err
	}

	paymentHash, err := crypto.GeneratePaymentHash(preimage)
	if err != nil {
		return nil, err
	}

	// Create hold invoice
	invoice, err := s.ln.AddHoldInvoice(req.AmountSats, paymentHash, req.Description, s.expiry)
	if err != nil {
		return nil, err
	}

	// Store escrow in repository
	escrow := &Escrow{
		ID:          paymentHash,
		PaymentHash: paymentHash,
		Preimage:    preimage,
		AmountSats:  req.AmountSats,
		Status:      StatusPending,
		Description: req.Description,
		Expiry:      s.expiry,
		CreatedAt:   time.Now().Unix(),
		UpdatedAt:   time.Now().Unix(),
	}

	if err := s.repo.Create(escrow); err != nil {
		return nil, err
	}

	return &CreateEscrowResponse{
		PaymentRequest: invoice.PaymentRequest,
		PaymentHash:    paymentHash,
		Preimage:       preimage,
		AmountSats:     req.AmountSats,
		Status:         string(StatusPending),
		Expiry:         s.expiry,
	}, nil
}

// ShipEscrowRequest represents the request to log shipment
type ShipEscrowRequest struct {
	PaymentHash    string `json:"payment_hash"`
	TrackingNumber string `json:"tracking_number"`
}

// ShipEscrow logs a shipment for an escrow
func (s *Service) ShipEscrow(ctx context.Context, req *ShipEscrowRequest) error {
	return s.repo.UpdateTrackingNumber(req.PaymentHash, req.TrackingNumber)
}

// PayEscrow simulates the buyer paying the invoice (funds become HELD)
func (s *Service) PayEscrow(ctx context.Context, paymentHash string) error {
	_, ok := s.repo.GetByPaymentHash(paymentHash)
	if !ok {
		return ErrEscrowNotFound
	}
	return s.repo.UpdateStatus(paymentHash, StatusHeld)
}

// ReleaseFunds releases funds to the merchant (called by oracle on delivery)
func (s *Service) ReleaseFunds(ctx context.Context, trackingNumber string) error {
	escrow, ok := s.repo.GetByTrackingNumber(trackingNumber)
	if !ok {
		return ErrEscrowNotFound
	}

	// Settle the invoice with the preimage
	if err := s.ln.SettleInvoice(escrow.Preimage); err != nil {
		return err
	}

	// Update status to settled
	return s.repo.UpdateStatus(escrow.PaymentHash, StatusSettled)
}

// CancelEscrow cancels an escrow and refunds the buyer
func (s *Service) CancelEscrow(ctx context.Context, paymentHash string) error {
	_, ok := s.repo.GetByPaymentHash(paymentHash)
	if !ok {
		return ErrEscrowNotFound
	}

	// Cancel the invoice
	if err := s.ln.CancelInvoice(paymentHash); err != nil {
		return err
	}

	// Update status to refunded
	return s.repo.UpdateStatus(paymentHash, StatusRefunded)
}

// GetEscrow retrieves an escrow by payment hash
func (s *Service) GetEscrow(ctx context.Context, paymentHash string) (*Escrow, error) {
	escrow, ok := s.repo.GetByPaymentHash(paymentHash)
	if !ok {
		return nil, ErrEscrowNotFound
	}
	return escrow, nil
}

// ListEscrows returns all escrows
func (s *Service) ListEscrows(ctx context.Context) []*Escrow {
	return s.repo.List()
}

File: internal/escrow/service.py
from __future__ import annotations

import time
from typing import Optional

from internal.escrow.repository import Escrow, EscrowNotFoundError, Repository
from internal.lightning.client import LightningClient
from pkg.crypto.random import generate_payment_hash, generate_preimage


class EscrowService:
    def __init__(self, repository: Repository, lightning_client: LightningClient, expiry: int) -> None:
        self.repository = repository
        self.lightning_client = lightning_client
        self.expiry = expiry

    def create_escrow(self, amount_sats: int, description: str) -> dict:
        preimage = generate_preimage()
        payment_hash = generate_payment_hash(preimage)

        invoice = self.lightning_client.create_hold_invoice(
            amount_sats=amount_sats,
            payment_hash=payment_hash,
            description=description,
            expiry=self.expiry,
        )

        escrow = Escrow(
            payment_hash=payment_hash,
            preimage=preimage,
            amount_sats=amount_sats,
            status="PENDING",
            description=description,
            expiry=self.expiry,
            created_at=int(time.time()),
            updated_at=int(time.time()),
        )
        self.repository.create(escrow)

        return {
            "payment_request": invoice["payment_request"],
            "payment_hash": payment_hash,
            "preimage": preimage,
            "amount_sats": amount_sats,
            "status": "PENDING",
            "expiry": self.expiry,
        }

    def ship_escrow(self, payment_hash: str, tracking_number: str) -> None:
        self.repository.update_tracking_number(payment_hash, tracking_number)

    def pay_escrow(self, payment_hash: str) -> None:
        self.repository.update_status(payment_hash, "HELD")

    def release_funds(self, tracking_number: str) -> None:
        escrow = self.repository.get_by_tracking_number(tracking_number)
        if escrow is None:
            raise EscrowNotFoundError("escrow not found")

        self.lightning_client.settle_invoice(escrow.preimage)
        self.repository.update_status(escrow.payment_hash, "SETTLED")

    def cancel_escrow(self, payment_hash: str) -> None:
        self.repository.update_status(payment_hash, "REFUNDED")
        self.lightning_client.cancel_invoice(payment_hash)

    def accept_escrow(self, payment_hash: str) -> None:
        escrow = self.repository.get_by_payment_hash(payment_hash)
        if escrow is None:
            raise EscrowNotFoundError("escrow not found")

        self.lightning_client.settle_invoice(escrow.preimage)
        self.repository.update_status(payment_hash, "SETTLED")

    def dispute_escrow(self, payment_hash: str) -> None:
        self.repository.update_status(payment_hash, "DISPUTED")

    def get_escrow(self, payment_hash: str) -> Optional[Escrow]:
        return self.repository.get_by_payment_hash(payment_hash)

    def list_escrows(self) -> list[Escrow]:
        return self.repository.list()
